### PR TITLE
feat: add yarn modern pnp support

### LINKED ISSUE #430: Error "Cannot find module 'cypress'"  using Yarn Berry
Hi there!
This action worked with yarn berry workspaces. but, this action can't find the "cypress" module.

```
Error: Cannot find module 'cypress'
```

I found the [issue](https://github.com/cypress-io/github-action/issues/170) of Yarn Berry and tried as follows.

- https://github.com/cypress-io/github-action/pull/260

here is the relavant configuration:

```
jobs:
  check-e2e-test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Use Node.js 14.x
        uses: actions/setup-node@v1
        with:
          node-version: 14.x

      - name: Cypress run
        uses: cypress-io/github-action@v2
        with:
          install-command : yarn install --immutable --immutable-cache
          start: yarn start
          wait-on: 'http://localhost:3000'
          browser: chrome
          headless: true
```

here is github-action log:

```
Run cypress-io/github-action@v2
  with:
    install-command: yarn install --immutable --immutable-cache
    build: yarn run build
    start: yarn start
    wait-on: http://localhost:3000
    browser: chrome
    headless: true
    record: false
    config-file: cypress.json
Received 22 of 22 (100.0%), 0.0 MBs/sec
Cache Size: ~0 MB (22 B)
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/be0c8559-6dda-42c1-822d-2db565488a1c/cache.tzst -P -C /home/runner/work/*****/*****
Cache restored successfully
Received 0 of 353031365 (0.0%), 0.0 MBs/sec
Received 58720256 of 353031365 (16.6%), 28.0 MBs/sec
Received 155189248 of 353031365 (44.0%), 49.3 MBs/sec
Received 251658240 of 353031365 (71.3%), 60.0 MBs/sec
Received 344642757 of 353031365 (97.6%), 65.7 MBs/sec
Received 353031365 of 353031365 (100.0%), 61.3 MBs/sec
Cache Size: ~337 MB (353031365 B)
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/dd9a16d5-2321-41d4-9f8f-3e8284d07e08/cache.tzst -P -C /home/runner/work/*****/*****
Cache restored successfully
install command command "yarn install --immutable --immutable-cache"
current working directory "/home/runner/work/*****/*****"
/usr/local/bin/yarn install --immutable --immutable-cache
➤ YN0000: ┌ Resolution step
Resolution step
➤ YN0000: └ Completed in 0s 376ms
➤ YN0000: ┌ Fetch step
Fetch step
➤ YN0000: └ Completed in 0s 743ms
➤ YN0000: ┌ Link step
Link step
➤ YN0000: └ Completed in 0s 397ms
➤ YN0000: Done with warnings in 1s 966ms
/opt/hostedtoolcache/node/14.17.6/x64/bin/npx cypress cache list
┌─────────┬───────────────────┐
│ version │ last used         │
├─────────┼───────────────────┤
│ 8.4.1   │ a few seconds ago │
├─────────┼───────────────────┤
│ 8.5.0   │ a few seconds ago │
└─────────┴───────────────────┘
build app command "yarn run build"
current working directory "/home/runner/work/*****/*****"
/usr/local/bin/yarn run build
The following changes are being made to your tsconfig.json file:
  - compilerOptions.paths must not be set (aliased imports are not supported)

Creating an optimized production build...
Compiled successfully.

File sizes after gzip:

  74.98 KB  build/static/js/2.c1fee284.chunk.js
  1.71 KB   build/static/js/3.b8a66e6b.chunk.js
  1.17 KB   build/static/js/runtime-main.2d193d74.js
  1.01 KB   build/static/js/main.a1fef262.chunk.js
  554 B     build/static/css/main.f09a1402.chunk.css

The project was built assuming it is hosted at /.
You can control this with the homepage field in your package.json.

The build folder is ready to be deployed.
You may serve it with a static server:

  yarn global add serve
  serve -s build

Find out more about deployment here:

  cra.link/deployment

start server "yarn start command "yarn start"
current working directory "/home/runner/work/*****/*****"
waiting on "http://localhost:3000" with timeout of 60 seconds
/usr/local/bin/yarn start
The following changes are being made to your tsconfig.json file:
  - compilerOptions.paths must not be set (aliased imports are not supported)

ℹ ｢wds｣: Project is running at http://10.1.0.5
ℹ ｢wds｣: webpack output is served from 
ℹ ｢wds｣: Content not from webpack is served from /home/runner/work/*****/*****/public
ℹ ｢wds｣: 404s will fallback to /
Starting the development server...

Compiled successfully!

You can now view ***** in the browser.

  Local:            http://localhost:3000
  On Your Network:  http://10.1.0.5:3000

Note that the development build is not optimized.
To create a production build, use yarn build.

Error: Cannot find module 'cypress'
Require stack:
- /home/runner/work/_actions/cypress-io/github-action/v2/dist/index.js
```

#### The above error occurred and was solved using with `command`.

```
  check-e2e-test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Use Node.js 14.x
        uses: actions/setup-node@v1
        with:
          node-version: 14.x

      - name: Cypress run
        uses: cypress-io/github-action@v2
        with:
          install-command : yarn install --immutable --immutable-cache
          command: yarn test:e2e
```

here is `yarn test:e2e` script:

```
"test:e2e": "start-server-and-test start http://localhost:3000 cypress:run",
```

Why can't I find the cypress module when I use the command provided by Cypress github-action? Or did I miss something?
I'd appreciate it if you had any other advice. thank you!🙏

### LINKED ISSUE #589: Error: Cannot find module 'cypress' after upgrade
After upgrading from 3 to 4 and Cypress 9 to 10, the following happens:

```yaml
      - name: Build frontend
        if: steps.reuse-frontend-build.outputs.cache-hit != 'true'
        uses: cypress-io/github-action@v4.2.0 # https://github.com/cypress-io/github-action
        env:
          DEBUG: "@cypress/github-action"
        with:
          working-directory: frontend
          install-command: yarn --immutable
          build: yarn build
          runTests: false
```

```
...
The build folder is ready to be deployed.
You may serve it with a static server:

  yarn global add serve
  serve -s build

Find out more about deployment here:

  https://cra.link/deployment

2022-08-01T19:30:07.309Z @cypress/github-action No start command found
2022-08-01T19:30:07.311Z @cypress/github-action Cannot find module 'cypress'
Require stack:
- /home/runner/work/_actions/cypress-io/github-action/v4.2.0/dist/index.js
2022-08-01T19:30:07.318Z @cypress/github-action Error: Cannot find module 'cypress'
Require stack:
- /home/runner/work/_actions/cypress-io/github-action/v4.2.0/dist/index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
Error: Cannot find module 'cypress'
Require stack:
- /home/runner/work/_actions/cypress-io/github-action/v4.2.0/dist/index.js
    at Function.resolve (node:internal/modules/cjs/helpers:108:19)
    at runTests (/home/runner/work/_actions/cypress-io/github-action/v4.2.0/dist/index.js:75237:13)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
0s
```

So it's clear to me that the action successfully installed and built the code, and then what is it trying to do, why must it do something more and fail at doing that? Before this upgrade, this CI script worked fine (with action v3 and cy 9).

BTW we're using Yarn 2 PnP. Starting to think about rolling back to a regular `node_modules` setup. This PnP thing made so many things so much more difficult, especially with Cypress when we first implemented it a year ago!

### PR DESCRIPTION
**Pull request by PilotConway**
Fixes the issue with users of yarn pnp where the require of cypress happens before the checks for command and command-prefix, causing the error `cannot find module 'cypress'`. 

yarn pnp users will be required to use the `command` property since those bypass the npx running of cypress and instead run through yarn which will import properly. 

fixes #589 
fixes #430 

* [NEW] Added examples of yarn pnp projects. 
* [CHG] Moved the cypress require statements to right before they are needed after the checks for command and command-prefix. 
* [CHG] Updated README with a note on running yarn pnp with command. 
* [CHG] build the dist files with the change.

### CODE REVIEW

**Comment by PilotConway** [comment]
Just updated because we found using `command-prefix` actually resulted in cypress being run via `npx yarn` which would fail when cypress versions were bumped. So changed the examples and documentation to suggest using command instead of command-prefix as that works more reliably in our tests.

**Comment by jribeiro** [comment]
Any update on this?

**Comment by MikeMcC399** [comment]
This really needs some feedback from @jaffrepaul and colleagues from the Cypress organization. The existing yarn examples in this repository are using [Yarn 1 (Classic)](https://classic.yarnpkg.com/lang/en/) and these should stay there because this version is still important.

Adding [Yarn Berry](https://github.com/yarnpkg/berry) (aka Yarn v2 & v3) examples would be a great addition.

**Comment by MikeMcC399** [comment]
@PilotConway 

My suggestion would be to separate out the changes to [index.js](https://github.com/cypress-io/github-action/blob/master/index.js) to fix the issues mentioned in the original post and put these changes into a separate PR. Possibly these could be fast-tracked. That depends on the capacity on the `cypress-io` team to review and merge.

The examples would then be a separate PR. Suggestion for the examples:

- drop Ubuntu 18 which is deprecated and add Ubuntu 22
- review whether it is necessary to add legacy v9 examples at this stage, or whether new examples should only be added for v10 config structure
- rebase on the current `master` branch
- use a branch name other than `master` for the source branch of the examples PR

I would also like to see a description in the README which makes it clearer about using the classic (v1) and current (Berry) Yarn versions. This might be in a separate PR though.

I wouldn't put any new effort into this until there is feedback from the repository owners.

**Comment by MikeMcC399** [comment]
@PilotConway 

I saw that you submitted some new commits, however you didn't write any comment. There hasn't been any comment from the Cypress.io core team either.

This comment should get your PR added to the support process.

**Comment by MikeMcC399** [comment]
@PilotConway 

I ran your branch and I found that the v9 job succeeds, whilst the v12 job fails.

## Suggestions

1. Put the PR in draft status until it runs correctly. You can check the results of running the action in your own fork.
2. Remove the v9 job and example. New v9 examples are no longer desired (see [CONTRIBUTING](https://github.com/cypress-io/github-action/blob/master/CONTRIBUTING.md#adding-a-new-example))
3. In the V12 example change the command to run `cypress` instead of `cypress@9`
4. Add a file `examples/yarn-modern-pnp/.gitignore`

    ```text
    !.yarn/cache
    !.pnp.*
    ```
5. Commit the `cache` and `pnp` files, otherwise you do not have a Yarn pnp environment set up.

## Comments

1. Cypress with Electron and Yarn pnp seems to have a compatibility issue. (see https://github.com/cypress-io/cypress/issues/19229 regarding the error "[ERROR:node_bindings.cc(279)] Most NODE_OPTIONs are not supported in packaged apps. See documentation for more details.")

2. Using the `command` parameter disables many other parameters. See [Custom test command](https://github.com/cypress-io/github-action#custom-test-command). So this may prevent its widespread use.

3. `yarn install` generates the message "ESM support for PnP uses the experimental loader API and is therefore experimental" so this also needs to be considered.

**Comment by PilotConway** [comment]
@MikeMcC399 Yeah, still working on a failing github action over in my repo after pulling in the latest. I was trying yesterday to see if your changes in v5 worked, but for PnP it still fails since it tried to import cypress and in pnp it's in a different location. I'm working this morning on getting the last action running then I'll update when it's ready again.

**Comment by MikeMcC399** [comment]
@PilotConway 

With the changes that I suggested, your PR is basically working. I didn't check if it causes compatibility issues with any of the other tests though.

**Comment by PilotConway** [comment]
@MikeMcC399 Thanks for those suggesstions. I'll remove that v9 then. 

As for the gitignore updates, adding the cache and pnp files are only required if you want to turn on zero installs. The bug is the same for regular pnp (where it downloads on yarn install) as it is with zero installs, so I think theres no need to add the additional files for any dependencies to the repo. The difference here is that removing `nodeLinker` in the yarn config is what causes the bug, since the action file can no longer find cypress.

**Comment by MikeMcC399** [comment]
@PilotConway 

> As for the gitignore updates, adding the cache and pnp files are only required if you want to turn on zero installs.

Ah, I'm a relative newcomer to yarn, so sorry if I didn't work out the difference between zero installs and pnp!

**Comment by PilotConway** [comment]
@MikeMcC399 No worries! It's definitely not simple getting it all setup.

This should be good to go now, the action is working on my fork and I ran the updated action to our internal application repo and it passed there as well.

File: README.md
Comment (mschile) [inline README.md:44]: Let's move this section above `Yarn workspaces`.

File: README.md
Comment (mschile) [inline README.md:1108]:
```suggestion
When using [Yarn Modern](https://yarnpkg.com/) (Yarn 2 and later) with [Plug'n'Play](https://yarnpkg.com/features/pnp) enabled, you will need to use the `command` parameter to run [yarn dlx](https://yarnpkg.com/cli/dlx) instead of [npx](https://docs.npmjs.com/cli/v9/commands/npx).
```

File: README.md
Comment (PilotConway) [reply README.md:44]: 🤦‍♂️ Yep, totally should have caught that!

**Review by mschile — APPROVED** [review]
LGTM, thanks for the contribution @PilotConway 🎉

**Comment by MikeMcC399** [comment]
@PilotConway 

I suggest to change the title to

```text
feat: add yarn modern pnp support
```

as it is more than just a fix. You are adding a new workflow with example code.

**Comment by MikeMcC399** [comment]
@PilotConway 

Congratulations on getting your first approval from the Cypress.io team!

This should help those who want to use [Yarn Plug'n'Play](https://yarnpkg.com/features/pnp), however the use of `command: yarn dlx cypress run` is really still a workaround. This PR should be followed up by further enhancements which allow [Yarn Plug'n'Play](https://yarnpkg.com/features/pnp) to work without using the `command` parameter. That would allow other features like the test summary or reporting of the Cypress Cloud link after recording to work.

**Comment by MikeMcC399** [comment]
Unfortunately there is an issue in Cypress with Yarn Modern PnP. See 

Running 

```bash
yarn
yarn dlx cypress open
```

followed by selecting E2E testing results in the Cypress app results in:

image

- A very similar issue is logged in https://github.com/cypress-io/cypress/issues/26567.

**Comment by PilotConway** [comment]
@MikeMcC399 Agree this is definitely a workaround to at least get it to work, and definitely think allowing the config to change `npx` to something else for all commands would be a huge help, but that's definitely a deeper change to the code when I was poking around. 

Also in terms of running locally, its been working for us locally without any issue, but I'm seeing we were a couple versions behind (12.8.1) and something broke recently in 12.10 so that would explain it. We don't use dlx locally and use the installed version directly. 

However, I think this is definitely a separate issue and the workaround can be attaching a version to the dlx command to run the version before this issue. `yarn dlx cypress@12.9` or `yarn add -D cypress@12.9 && yarn cypress` to run the UI. I might try a few things over the next week and see if I dig up something, but thanks for pointing me to that issue, I'll subscribe to that post if I find anything on our end.

**Review by chrisbreiding — APPROVED** [review]

**Comment by MikeMcC399** [comment]
It seems that

```bash
yarn add bluebird@3.7.2 -D -E
```

will work around the issue for E2E testing.

**Comment by MikeMcC399** [comment]
@PilotConway 

I suggest to take `dlx` out of the example. When it is used without a version number it uses the latest version, which is not necessarily the one defined in `yarn.lock`. It would be possible to add a version number, however that makes maintenance more cumbersome, as the version would need to be changed in two places when doing an update.

I tried out using just

```bash
yarn cypress run
```

and that looks to work well. It starts the version of Cypress already installed in the `.yarn` directory, which would be the one required.

What do you think?

The [Yarn dlx](https://yarnpkg.com/cli/dlx) documentation does not explain this well. I just tried it out and this is the way it works with Yarn `3.5.1`.

**Comment by MikeMcC399** [comment]
- I have submitted a draft PR https://github.com/cypress-io/github-action/pull/903 in order to make a concrete suggestion for this change

**Pull request merged**
